pcsx2/gui/Panels/BiosSelectorPanel.cpp

```diff
@@ -162,28 +167,48 @@ bool Panels::BiosSelectorPanel::ValidateEnumerationStatus()
 	return validated;
 }
 
-void Panels::BiosSelectorPanel::DoRefresh()
+void Panels::BiosSelectorPanel::EnumThread::ExecuteTaskInThread()
 {
-	if (!m_BiosList) return;
+	for (size_t i = 0; i < m_parent.m_BiosList->GetCount(); ++i)
+	{
+		wxString description;
+		if (!IsBIOS((*m_parent.m_BiosList)[i], description)) continue;
+		Result.emplace_back(std::make_pair(description, i));
+	}
+
+	wxCommandEvent done(pxEvt_BiosEnumerationFinished);
+	done.SetClientData(this);
+	m_parent.GetEventHandler()->AddPendingEvent(done);
+}
 
+void Panels::BiosSelectorPanel::DoRefresh()
+{
 	m_ComboBox->Clear();
+	if (!m_BiosList->size()) return;
 
-	const wxFileName right(g_Conf->FullpathToBios());
-	bool biosSet = false;
+	m_ComboBox->Append(wxString("Enumerating BIOSes..."));
+	m_ComboBox->Update();
 
-	for(size_t i=0; i<m_BiosList->GetCount(); ++i)
-	{
-		wxString description;
-		if (!IsBIOS((*m_BiosList)[i], description)) continue;
-		int sel = m_ComboBox->Append( description, (void*)i );
+	m_EnumeratorThread.reset(new EnumThread(*this));
+
+	m_EnumeratorThread->Start();
+}
+
+void Panels::BiosSelectorPanel::OnEnumComplete(wxCommandEvent &evt)
+{
+	auto enumThread = static_cast<EnumThread*>(evt.GetClientData());
+	// Sanity check, in case m_BiosList was updated by ValidateEnumerationStatus() while the EnumThread was running
+	if (m_EnumeratorThread.get() != enumThread || m_BiosList->size() < enumThread->Result.size()) return;
+
+	const wxFileName& currentBios = g_Conf->FullpathToBios();
+	m_ComboBox->Clear(); // Clear the "Enumerating BIOSes..."
 
-		if (wxFileName((*m_BiosList)[i] ) == right)
-		{
+	std::for_each(enumThread->Result.begin(), enumThread->Result.end(), [this, currentBios](std::pair<wxString, u32> result) {
```

Comment:
```suggestion
	for (const std::pair<wxString, u32>& result : enumThread->Result)
	{
```
I know everyone loves their higher order functions, but C++ does a really bad job of making them look nice